### ISSUE TITLE
Behaviour of envi dataset and geoTransform

### ISSUE THREAD

**Issue by jmichel-otb** [OSGeo/gdal#944]
https://github.com/OSGeo/gdal/blob/05323c6f6b81ecbdcc78e2305e9c62e1f52a2d41/gdal/frmts/raw/envidataset.cpp#L621

I have a weird behaviour which I can trace back to this part of the envidataset code:

I am trying to write a dataset with the following geoTransform : {-0.5, 1, 0, -0.5, 0, 1}. 
I end up with:
- A weird warning (with `CPL_DEBUG=ON`): 
```
ENVI: rot1 = 0, rot2 = 180
Warning 1: Geotransform matrix has non rotational terms
```
- A hdr file with the rotated keyword in the map_projection info:
`map info = {Arbitrary, 1, 1, -0.5, -0.5, 1, 1, 0, North, rotation=90}` 
- A weird geoTransform when trying to read back the dataset with gdal: 
```
GeoTransform =
  -0.5, 6.123233995736766e-17, 1
  -0.5, 1, -6.123233995736766e-17
```
It is weird that the two atan2 calls gives 0 and 180 respectively, which triggers the warning. In particular I an not sure about the second call:
`            -atan2(-adfGeoTransform[4], -adfGeoTransform[5]) * kdfRadToDeg;
`
I do not understand the reason for `-adfGeoTransform[5]` ?

**Comment by rouault** [comment]
> I do not understand the reason for -adfGeoTransform[5] ?

This comes from the "forward path" (coming from ENVI convention to GDAL geotransform)
```
    adfGeoTransform[0] = pixelEasting - (xReference - 1) * xPixelSize;
    adfGeoTransform[1] = cos(dfRotation) * xPixelSize;
    adfGeoTransform[2] = -sin(dfRotation) * xPixelSize;
    adfGeoTransform[3] = pixelNorthing + (yReference - 1) * yPixelSize;
    adfGeoTransform[4] = -sin(dfRotation) * yPixelSize;
    adfGeoTransform[5] = -cos(dfRotation) * yPixelSize;
```
I guess the issue is that the ENVI format might not be able to represent non-georeferenced image like what you try to do, at least not the way GDAL tries to encode it.

**Comment by jmichel-otb** [comment]
Ok, I guess the best we can do is to avoid using ENVI file format ... In gdal tutorials, I read:
`adfGeoTransform[5] /* n-s pixel resolution (negative value) */`
Is this a requirement of gdal API ? I am under the impression that this is assumed by envidataset code, and that it causes the problem.

Another question is the warning : Gdal says that the geoTransform has non rotational terms but encodes a (wrong) rotation anyway. Maybe a better behaviour would be to warn but not encode in that case ?

**Comment by rouault** [comment]
> Is this a requirement of gdal API ?

No, this is a general hint. For georeferenced case, this is a negative value ?

> Gdal says that the geoTransform has non rotational terms but encodes a (wrong) rotation anyway. Maybe a better behaviour would be to warn but not encode in that case ?

I've no strong opinion there. Perhaps there would be situations with slight shearing and people would still want some form of georeferencing to be written. A warning is emitted so users should be careful...  
Clearly in your case, the difference between the 2 computed angles is so large that an error would be more appropriate

**Comment by jmichel-otb** [comment]
Thanks @rouault, there are still two things that I do not understand:

Class constructor sets geotransform to:
```
    adfGeoTransform[0] = 0.0;
    adfGeoTransform[1] = 1.0;
    adfGeoTransform[2] = 0.0;
    adfGeoTransform[3] = 0.0;
    adfGeoTransform[4] = 0.0;
    adfGeoTransform[5] = 1.0;
```
Which is likely to trigger the warning and the wrong write/read back of geotransform.

I investigate this issue because our code which writes hdr files with `geoTransform[5]>0` using gdal broke at some point (which means that the same code once worked ...). I can not figure out why. History of `envidataset.cpp` shows that the code is like this from the beginning.

**Comment by grizonnetm** [comment]
Just want to add that I've made some test to reproduce the problem reported by @jmichel-otb which pop up in one OTB test (https://dash.orfeo-toolbox.org/testDetails.php?test=87036007&build=336029) 

I check the test status using OTB build against different versions of GDAL. The test pass with GDAL version 2.2.1 which is the current version of GDAL packaged in OTB standalone binaries but fails with GDAL 2.2.3 (version available on Ubuntu 18.04).

**Comment by jmichel-otb** [comment]
Thank you @rouault, I suspect this was masking the problem, and that the introduction of rotated datasets support in envi driver actually unveiled it. Only option we have is to:

- Change format of hdr files in our testing dataset to something else (tif probably)
- Warn (or except) in OTB if someone tries to write to envi file format with a postive y spacing.
